pilot/pkg/networking/grpcgen/cds.go

```diff
@@ -0,0 +1,270 @@
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/core/v1alpha3"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pilot/pkg/util/sets"
+	"istio.io/istio/pkg/config/host"
+)
+
+// BuildClusters handles a gRPC CDS request, used with the 'ApiListener' style of requests.
+// The main difference is that the request includes Resources to filter.
+func (g *GrpcConfigGenerator) BuildClusters(node *model.Proxy, push *model.PushContext, names []string) model.Resources {
+	filter := newClusterFilter(names)
+	var clusters = make([]*clusterv3.Cluster, 0, len(names))
+	for defaultClusterName, subsetFilter := range filter {
+		builder, err := newClusterBuilder(node, push, defaultClusterName, subsetFilter)
+		if err != nil {
+			log.Warn(err)
+			continue
+		}
+		clusters = append(clusters, builder.build()...)
+	}
+
+	resp := make(model.Resources, 0, len(clusters))
+	for _, c := range clusters {
+		resp = append(resp, &envoy_service_discovery_v3.Resource{
+			Name:     c.Name,
+			Resource: util.MessageToAny(c),
+		})
+	}
+	return resp
+}
+
+// newClusterFilter builds a filtering map to determine which clusters need to be built.
+// gRPC will usually request each subset individually, regardless of if a previous response included it.
+func newClusterFilter(names []string) map[string]sets.Set {
+	filter := map[string]sets.Set{}
+	for _, name := range names {
+		dir, _, hn, p := model.ParseSubsetKey(name)
+		defaultKey := model.BuildSubsetKey(dir, "", hn, p)
+		if _, ok := filter[defaultKey]; !ok {
+			filter[defaultKey] = sets.NewSet()
+		}
+		filter[defaultKey].Insert(name)
+	}
+	return filter
+}
+
+// clusterBuilder is responsible for building a single default and subset clusters for a service
+// TODO re-use the v1alpha3.ClusterBuilder:
+// Most of the logic is similar, I think we can just share the code if we expose:
+// * BuildSubsetCluster
+// * BuildDefaultCluster
+// * BuildClusterOpts and members
+// * Add something to allow us to override how tlscontext is built
+type clusterBuilder struct {
+	// conveinence
+	push *model.PushContext
+	node *model.Proxy
+
+	// guaranteed to be set in init
+	defaultClusterName   string
+	requestedClusterName string
+	hostname             host.Name
+	portNum              int
+
+	// may not be set
+	svc    *model.Service
+	port   *model.Port
+	filter sets.Set
+}
+
+func newClusterBuilder(node *model.Proxy, push *model.PushContext, defaultClusterName string, filter sets.Set) (*clusterBuilder, error) {
+	_, _, hostname, portNum := model.ParseSubsetKey(defaultClusterName)
+	if hostname == "" || portNum == 0 {
+		return nil, fmt.Errorf("failed parsing subset key: %s", defaultClusterName)
+	}
+
+	// try to resolve the service and port
+	var port *model.Port
+	svc := push.ServiceForHostname(node, hostname)
+	if svc == nil {
+		log.Warnf("cds gen for %s: did not find service for cluster %s", node.ID, defaultClusterName)
+	} else {
+		var ok bool
+		port, ok = svc.Ports.GetByPort(portNum)
```

Comment:
imo this is cleaner
```suggestion
		port, ok := svc.Ports.GetByPort(portNum)
```

Reply:
`port` needs to be in the outer scope, so I could do 

```
svcPort, ok := ...
...
port = svcPort
```

if that feels cleaner?

Reply:
nevermind, what I was trying to do isnt possible. Your way is good